system-addon/lib/TopStoriesFeed.jsm

```diff
@@ -277,6 +298,81 @@ this.TopStoriesFeed = class TopStoriesFeed {
     this.newTabsSinceSpoc++;
   }
 
+  // Frequency caps are based on campaigns, which may include multiple spocs.
+  // We currently support two types of frequency caps:
+  // - lifetime: Indicates how many times spocs from a campaign can be shown in total
+  // - period: Indicates how many times spocs from a campaign can be shown within a period
+  //
+  // So, for example, the feed configuration below defines that for campaign 1 no more
+  // than 5 spocs can be show in total, and no more than 2 per hour.
+  // "campaign_id": 1,
+  // "caps": {
+  //  "lifetime": 5,
+  //  "campaign": {
+  //    "count": 2,
+  //    "period": 3600
+  //  }
+  // }
+  isBelowFrequencyCap(impressions, spoc) {
+    const campaignImpressions = impressions[spoc.spoc_meta.campaign_id];
+    if (!campaignImpressions) {
+      return true;
+    }
+
+    const lifeTimeCap = spoc.spoc_meta.caps && spoc.spoc_meta.caps.lifetime;
+    const lifeTimeCapExceeded = campaignImpressions.length >= lifeTimeCap;
```

Comment:
Given that `lifeTimeCap` and `campaignCap` are both from 3rd party, we might want to have a builtin cap defined by us, which prevents the timestamp array from growing without bound. Just be safe :)